### PR TITLE
Local chop for CompositionalMultiphaseFVM

### PR DESCRIPTION
**Pull request by paveltomin**
Rebase PR: https://github.com/GEOS-DEV/integratedTests/pull/44

### CODE REVIEW

File: src/coreComponents/physicsSolvers/SolverBase.hpp [hunk 1]
```diff
@@ -784,6 +785,9 @@ class SolverBase : public ExecutableGroup
   /// Nonlinear solver parameters
   NonlinearSolverParameters m_nonlinearSolverParameters;
 
+  /// Flag for local chop
+  integer m_localChop = 0;
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/SolverBase.hpp:789]: should it be here or in NonlinearSolverParameters?

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp [hunk 2]
```diff
@@ -442,13 +442,13 @@ real64 CompositionalMultiphaseHybridFVM::scalingForSystemSolution( DomainPartiti
...
-                                                                MeshLevel const & mesh,
+                                                                MeshLevel & mesh,
                                                                 arrayView1d< string const > const & regionNames )
   {
     mesh.getElemManager().forElementSubRegions< ElementSubRegionBase >( regionNames, [&]( localIndex const,
-                                                                                          ElementSubRegionBase const & subRegion )
+                                                                                          ElementSubRegionBase & subRegion )
     {
-      real64 const subRegionScalingFactor =
+      auto const subRegionData =
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp:451]: will remove auto later

File: src/coreComponents/physicsSolvers/fluidFlow/IsothermalCompositionalMultiphaseBaseKernels.hpp [hunk 3]
```diff
@@ -947,11 +984,28 @@ class ScalingAndCheckingSystemSolutionKernelBase
    */
   struct StackVariables
   {
+    StackVariables() {}
+
+    StackVariables( TYPE _localMinVal, real64 _localMaxDP, real64 _localMaxDCompDens,
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/IsothermalCompositionalMultiphaseBaseKernels.hpp:989]: any way to make it more elegant?

**Comment by paveltomin** [comment]
I also added some useful output for max DP, DC, DS to better observe the nonlinear behavior, only for logLevel >= 1

File: src/coreComponents/physicsSolvers/SolverBase.hpp (hunk 1 above)
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/SolverBase.hpp:789]: Yes, I would move it to `NonlinearSolverParameters`. For the naming, we can maybe stick to "scaling" to follow the nomenclature used up to now in GEOS (for instance in the function `scalingForSystemSolution`).

If you agree, in `NonlinearSolverParameters`, we could have something like:
```
enum class ScalingType : integer
{
    Global, ///< Scale the Newton update with a unique scaling factor
    Local    ///< Scale the Newton update locally (modifies the Newton direction)
};
```

**Review by francoishamon — COMMENTED** [review]
Looks very good so far

**Review by francoishamon — COMMENTED** [review]
Looks very good so far

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp [hunk 4]
```diff
@@ -508,24 +512,23 @@ void CompositionalMultiphaseBase::updatePhaseVolumeFraction( ObjectManagerBase &
...
+:    isothermalCompositionalMultiphaseBaseKernels::
+                   PhaseVolumeFractionKernelFactory::
+                   createAndLaunch< parallelDevicePolicy<> >( m_numComponents,
+                                                              m_numPhases,
+                                                              dataGroup,
+                                                              fluid );
+
+  maxDS = MpiWrapper::max( maxDS );
+  if( maxDS > 0 )
+    GEOS_LOG_LEVEL_RANK_0( 1, getName() << ": Max DS = "<< maxDS << " in subregion " << dataGroup.getName() );
```
[the remark this answers is not in the document]
Comment (paveltomin) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:531]: Sounds good idea, I was not aware of too many subRegions cases.

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp [hunk 5]
```diff
@@ -387,14 +389,45 @@ real64 CompositionalMultiphaseFVM::scalingForSystemSolution( DomainPartition con
...
+  maxDCompDens = MpiWrapper::max( maxDCompDens );
+  maxDT = MpiWrapper::max( maxDT );
+  minPresScalFac = MpiWrapper::min( minPresScalFac );
+  minCompDensScalFac = MpiWrapper::min( minCompDensScalFac );
+  minTempScalFac = MpiWrapper::min( minTempScalFac );
+
+  if( m_isThermal )
+  {
+    GEOS_LOG_LEVEL_RANK_0( 1, getName() + ": Max dP = " << maxDP << ", Max dCompDens = " << maxDCompDens << ", Max dT = " << maxDT << " (before scaling)" );
+  }
```
[the remark this answers is not in the document]
Comment (paveltomin) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp:414]: logLevel 2 starts printing matricies I think, not that much room for levels to be honest

**Comment by paveltomin** [comment]
Just realized few things for this PR:
1. My implementation is not compatible with line search - I added a check for that and I think it is ok for now.
2. checkSystemSolution needs to be also adjusted or the logic should be changed - update first then check. @francoishamon, what do you think?

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp [hunk 6]
```diff
@@ -361,16 +361,18 @@ real64 CompositionalMultiphaseFVM::scalingForSystemSolution( DomainPartition con
 
   string const dofKey = dofManager.getKey( viewKeyStruct::elemDofFieldString() );
   real64 scalingFactor = 1.0;
+  real64 maxDeltaPres = 0.0, maxDeltaCompDens = 0.0, maxDeltaTemp = 0.0;
```
Comment (dkachuma) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp:364]: I can't quite tell if the locally calculated pressure and temperature changes are absolute. If not I would expect `maxDeltaPres` and `maxDeltaTemp` to start as large negative values.

Comment (paveltomin) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp:364]:
```
    // compute the change in pressure
    real64 const pres = m_pressure[ei];
    real64 const absPresChange = LvArray::math::abs( m_localSolution[stack.localRow] );
    if( stack.localMaxDeltaPres < absPresChange )
    {
      stack.localMaxDeltaPres = absPresChange;
    }
```
this is how the change is computed, so it should be absolute

File: src/coreComponents/physicsSolvers/NonlinearSolverParameters.cpp [hunk 7]
```diff
@@ -157,6 +162,10 @@ void NonlinearSolverParameters::postProcessInput()
   {
     GEOS_ERROR( " timeStepIncreaseIterLimit should be smaller than timeStepDecreaseIterLimit!!" );
```
Comment (dkachuma) [inline src/coreComponents/physicsSolvers/NonlinearSolverParameters.cpp:163]: I know it's not part of this PR, but can we take the opportunity to remove the double exclamation mark here. I find it frightening.

File: src/coreComponents/physicsSolvers/fluidFlow/IsothermalCompositionalMultiphaseBaseKernels.hpp [hunk 8]
```diff
@@ -1240,28 +1415,39 @@ class SolutionCheckKernel : public ScalingAndCheckingSystemSolutionKernelBase< i
    * @param[in] compDens the component density vector
    */
   SolutionCheckKernel( integer const allowCompDensChopping,
+                       integer const localScaling,
```
Comment (dkachuma) [inline src/coreComponents/physicsSolvers/fluidFlow/IsothermalCompositionalMultiphaseBaseKernels.hpp:1418]: If possible can we propagate `NonlinearSolverParameters::ScalingType` just to highlight the values expected by this parameter?

**Review by dkachuma — APPROVED** [review]
Looks good to me

File: src/coreComponents/physicsSolvers/fluidFlow/IsothermalCompositionalMultiphaseBaseKernels.hpp (hunk 8 above)
Comment (paveltomin) [reply src/coreComponents/physicsSolvers/fluidFlow/IsothermalCompositionalMultiphaseBaseKernels.hpp]: Modified, please have a look

**Comment by paveltomin** [comment]
> Just realized few things for this PR:
> 
> 1. My implementation is not compatible with line search - I added a check for that and I think it is ok for now.
> 2. checkSystemSolution needs to be also adjusted or the logic should be changed - update first then check. @francoishamon, what do you think?

Actually, checkSystemSolution is probably ok, it is already adjusted at flow solver level to support local scaling

**Pull request merged**